cpp/src/utilities/cython.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithms.hpp>
+#include <utilities/cython.hpp>
+
+#include <raft/handle.hpp>
+
+namespace cugraph {
+namespace cython {
+
+// Factory function for creating graph containers from basic types
+// FIXME: This should accept void* for offsets and indices as well and take a
+//        dtype directly for each instead of the enum/int.
+graph_container_t create_graph_t(raft::handle_t const& handle,
+                                 int* offsets,
+                                 int* indices,
+                                 void* weights,
+                                 int weightType,
+                                 int num_vertices,
+                                 int num_edges,
+                                 int* local_vertices,
+                                 int* local_edges,
+                                 int* local_offsets,
+                                 bool transposed,
+                                 bool multi_gpu)
+{
+  graph_container_t graph_container{};
+  graph_container.wType = weightType;
+
+  if (weightType == floatType) {
+    graph_container.graph.GraphCSRViewFloat = GraphCSRView<int, int, float>(
+      offsets, indices, reinterpret_cast<float*>(weights), num_vertices, num_edges);
+    graph_container.graph.GraphCSRViewFloat.set_local_data(
+      local_vertices, local_edges, local_offsets);
+    graph_container.graph.GraphCSRViewFloat.set_handle(const_cast<raft::handle_t*>(&handle));
+
+  } else {
+    graph_container.graph.GraphCSRViewDouble = GraphCSRView<int, int, double>(
+      offsets, indices, reinterpret_cast<double*>(weights), num_vertices, num_edges);
+    graph_container.graph.GraphCSRViewDouble.set_local_data(
+      local_vertices, local_edges, local_offsets);
+    graph_container.graph.GraphCSRViewDouble.set_handle(const_cast<raft::handle_t*>(&handle));
+  }
+
+  return std::move(graph_container);
```

Comment:
Please return-by-value,
```
return graph_container;
```
 rather than convert to r-value reference. This cast (`std::move()`) prevents the compiler to perform the very powerful return-by-value optimization (RVO), which allows elimination of copying (_copy elision_).

RVO, or copy elision allows the compiler to allocate (under a few typical circumstance, like the one here; i.e., local variable returned without any conversion) the local returned variable directly into the caller's variable that gets assigned to the result of calling the function.

Reply:
done in latest commit